Reaction Rate Tallies, Documentation, Docstrings, PEP8

wbinventor opened the pull request:
This PR adds reaction rate mesh tally feature to the `openmoc.process` module. The goal for this is to make it easy to compare pin-by-pin or assembly-by-assembly reaction rates (nuclide-integrated fission, U-238 capture, etc.) between OpenMOC and other codes (*e.g.*, OpenMC).

In addition, this PR addresses issue #60 by introducing the widely used ["doxy2swig.py"](https://github.com/m7thon/doxy2swig) script to "docs/Doxygen" to convert our Doxygen comments in the C/C++/CUDA source code into docstrings embedded in the SWIG-generated Python module. It is *very* helpful to have docstrings for rapid interactive prototyping in ipython. Similarly, all of the Python sub-modules have been revised to use pure Python docstrings instead of Doxygen comments. 

Finally, the majority of this code changes in this PR are dedicated to new tests for `openmoc.process` and updated documentation on `openmoc.materialize`, `openmoc.plotter` and `openmoc.process` modules to address sub-issues in #194.

All of the changes made in this PR are summarized below:

* Added `Mesh` class to `openmoc.process` module
* Added "doxy2swig.py" to inject Doxygen comments as docstrings in SWIG-generated `openmoc` module
* Revised all Python modules to be (closer to) PEP8 conformant  (*i.e.*, 4 spaces / tab)
* Turned all Python module Doxygen comments into docstrings
* Updated documentation on `openmoc.materialize` module
* Updated docs on `openmoc.plotter` module
* Updated docs on `openmoc.process` module
* Updated docs on `openmoc.log` module
* Renamed `Circle` as `ZCylinder` in the docs
* Removed deprecated ``-no-numpy`` compile time option from docs
* Revised `Options` class to use Python properties instead of getter methods
* Revised error checking in Python modules to use new `openmc.checkvalue` module
* Added test for `openmoc.process` store/restore simulation states
* Added test for `openmoc.process` mesh tallies
* Added test for `openmoc.log` module
* Fixed bug in `Solver::getFSRSource(...)` method
* Removed unsupported `openmoc.compatible.casmo` module

samuelshaner commented on docs/source/usersguide/input.rst:
```diff
@@ -172,71 +169,63 @@ OpenMOC uses multi-group macroscopic nuclear cross-sections, provided by the use
    ...
 
-For many simulations, defining the nuclear data cross-sections by hand in a Python script is cumbersome and error-prone. As a result, OpenMOC includes the ``openmoc.materialize`` module for importing nuclear data cross-sections from an HDF5_ or a Python pickle_ binary file. The ``materialize(...)`` routine is used to import data and instantiate ``Material`` objects returned via a Python dictionar ...
+For many simulations, defining the nuclear data cross sections by hand in a Python script is cumbersome and error-prone. As a result, OpenMOC includes the ``openmoc.materialize`` module for importing nuclear data cross sections from an HDF5_ binary file. The ``load_from_hdf5(...)`` routine is used to import data and instantiate ``Material`` objects returned via a Python dictionary_. The use of th ...
```
illusrated -> illustrated

wbinventor replied:
good catch

samuelshaner commented on docs/source/usersguide/input.rst:
```diff
@@ -172,71 +169,63 @@ OpenMOC uses multi-group macroscopic nuclear cross-sections, provided by the use
-
-    # Retrieve the material called 'fuel' in the pickle file
-    fuel = pickle_materials['fuel']
+The ``openmoc.materialize`` module defines a standard for cross section data stored in binary files. First, HDF5 files must include an ``'# groups'`` attribute with the integer number of groups in the top level of the file hierarchy. Second, the string domain type - ``'material'`` or ``'cell'`` - must be specified in the top level of the file hierarchy. This must match the ``domain_type`` keyword ...
```
I feel like `include a ``# groups`` attribute` would be more appropriate.

wbinventor replied:
agreed

samuelshaner commented on docs/source/usersguide/input.rst:
```diff
@@ -811,7 +790,7 @@ OpenMOC has an integrated CMFD acceleration framework that allows users to great
     cmfd.setOpticallyThick(False)
     cmfd.setSORRelaxationFactor(1.5)
     cmfd.setMOCRelaxationFactor(0.66)
```
Can you remove this line since we no longer have an MOC relaxation factor?

wbinventor replied:
got it

wbinventor replied:
I also removed a full paragraph description of `Cell:setMOCRelaxationFactor()` below.

samuelshaner replied:
Great, thanks!

samuelshaner commented on docs/source/usersguide/input.rst:
```diff
@@ -826,7 +805,7 @@ These lines of code should be placed in your input file at the location where th
   * ``setGroupStructure`` (default: same as MOC group structure) - OpenMOC is able to perform CMFD on a coarse energy group structure to allow fine energy group problems to be accelerated with CMFD without incurring a significant computational overhead for CMFD. This function takes a python list as input with the first value of 1 (to indicate the first energy group) followed by an increasing valu ...
   * ``setOpticallyThick`` (default: False) - OpenMOC uses an correction factor on the material diffusion coefficients as described in the Theory and Methodology section. This correction factor is turned off by default.
```
While you are editing this file can you change `an` to `a`?

wbinventor replied:
got it

samuelshaner commented on docs/source/usersguide/processing.rst:
```diff
@@ -103,18 +103,119 @@ The code snippet below illustrates one possible configuration of parameters to t
+========================  ===================  ==========  ==========================================================
+``dimension``             list of integers     None        The number of mesh cells in each direction
+``lower_left``            list of floats       None        The lower-left corner of the structured mesh
+``upper_right``           list of floats       None        The upper-right corner of the structrued mesh
```
Do we want to go with `lower_left` and `upper_right` or `origin` and `offset` as is done with a lattice?

wbinventor replied:
The reason I went with `lower_left` and `upper_right` was to mirror the `openmc.Mesh` class as much as possible. At least from my perspective, this mesh tally feature is intended to make it very easy to compare spatially-integrated reaction rates between the two codes. The way it is implemented now, it is very easy to create an `openmoc.process.Mesh` from an `openmc.mesh`:

```python
...

# Extract mesh for mesh tallies from OpenMC StatePoint
openmc_mesh = statepoint.meshes[mesh_id]

# Create OpenMOC Mesh to mirror OpenMC Mesh
openmoc_mesh = openmoc.process.Mesh()
openmoc_mesh.dimension = openmc_mesh.dimension
openmoc_mesh.lower_left = openmc_mesh.lower_left
openmoc_mesh.upper_right = openmc_mesh.upper_right
openmoc_mesh.width = openmc_mesh.width

...
```

I think we should eitheradopt a standard akin to our own `Lattice` or `openmc.Mesh`. Based on the use case, I would go with the latter since it simplifies the code above (though I recognize that `lower_left` and `upper_right` can easily be backed out of `origin` and `offset` and vice versa).

samuelshaner replied:
Ok, I'm on board with that.

samuelshaner commented on src/CPUSolver.cpp:
```diff
@@ -889,12 +889,12 @@ void CPUSolver::computeFSRFissionRates(double* fission_rates, int num_FSRs) {
     fission_rates[r] = 0.0;
 
   /* Loop over all FSRs and compute the volume-averaged nu-fission rate */
```
nu-fission -> fission

wbinventor replied:
Haha, yes indeed. I changed it here as well as in `GPUSolver`

samuelshaner commented:
Could the `tally_fission_rates` method just create a `domains_to_coeffs` dict and call the `tally_on_mesh` method? I feel like that would shorten it up a bit.

samuelshaner commented:
Line 900 of `process.py` looks like it was copied from the doc string for the `tally_fission_rates` method and should be updated for the `tally_on_mesh` method.

samuelshaner commented:
@wbinventor this PR looks great! I'm really excited to use the new mesh tally features!

wbinventor commented:
@samuelshaner I addressed all of your comments. In addition, I added a new parent/ancestor feature to the `Cell` class to facilitate easier comparison of reaction rates between non-discretized OpenMC meshes and discretized OpenMOC meshes.

As I've gone about trying to compare various nuclide-specific reaction rates between OpenMC and OpenMOC, I've realized that it isn't as straightforward as the nuclide-integrated fission rates. The reason is that an OpenMC MGXS `Library` is defined for the `Cells` or `Materials` used to create the OpenMOC `Geometry` *prior* to discretization. Once discretized, the `Cells` and `Materials` in the `Geometry` do not easily map back to those in the `Library`. By tracking a *parent `Cells`* when a `Cell` is cloned during spatial discretization, we can easily keep retain a mapping from pre- to post-discretized geometries.

wbinventor commented:
I'm unsure why the test suite continues to fail, but will figure this out next.

samuelshaner commented:
@wbinventor I left two comments in the conversations of this PR that address some items in `process.py` that it appears you might have missed. I wasn't able to have the comment address a particular line because the diff was too large.

samuelshaner commented:
I'm curious, is there any chance the parent cell could be garbage collected?

wbinventor commented:
@samuelshaner - I fixed the copy-paste docstring issue. I had considered making `tally_fission_rates` use the more general `tally_on_mesh(...)` method but decided against it for two reasons. First, the process of constructing a `domains_to_coeffs` dictionary actually increases the amount of code in the method overall. Second, we already have the `Solver::computeFSRFissionRates(...)` method implemented in C++/CUDA which is faster than computing each of the rates fsr-by-fsr and group-by-group in Python. This second reason is a minor one since the computation is so fast anyway, but `tally_fission_rates` is noticeably faster than `tally_on_mesh` for some of the problems I've looked at.

wbinventor commented:
I don't think that the parent `Cell` can be garbage collected. Its `thisown` should be set to `False` once it is assigned to a `Universe`, meaning that we have actually allowed these parent `Cells` to be a memory leak (*i.e.*, the user can still manipulate the parent `Cell`, but neither Python nor C++ will delete its memory). I didn't fix this memory leak here since 1) it's not clear how to do so and 2) it is so little memory that it likely will never matter.

wbinventor commented:
I think the test suite issues have been fixed.

samuelshaner commented:
Ok, I think this is about ready to merge. Do you know why `test_simulation_state.py` is failing for Python 3?

wbinventor commented:
Yeah, I think I just fixed the Python 3 issue. Evidently Python 3 reports more decimal places for `float` variables than does Python 2.

I also re-ran Doxygen and "doxy2swig.py" to reflect the new parent `Cell` feature. @samuelshaner @geogunow - you should keep in mind that all future PRs should only be merged in after running Doxgyen and "doxy2swig.py" to update "openmoc/docstring.i" with docstrings synchronized to the C++ source code. This is akin to having to update the XML schema and statepoint revision for every PR to OpenMC.
